Fix incorrect merge suppressing my change to publish CG2 framework

trylek opened the pull request:
As David discovered, I apparently made a merge bug in rebasing my
CG2 framework switch-over change prior to checking in that actually
suppressed CG2 publishing due to losing the property
PublishReadyToRunUseCrossgen2. This change fixes that deficiency.

Thanks

Tomas

/cc @dotnet/crossgen-contrib

trylek committed:
Fix incorrect merge suppressing my change to publish CG2 framework

As David discovered, I apparently made a merge bug in rebasing my
CG2 framework switch-over change prior to checking in that actually
suppressed CG2 publishing due to losing the property
PublishReadyToRunUseCrossgen2. This change fixes that deficiency.

Thanks

Tomas

trylek committed:
Fix DOTNET_ROOT location

davidwrighton commented on src/installer/pkg/sfx/Microsoft.NETCore.App/ReadyToRun.targets:
```diff
@@ -48,7 +49,7 @@
     <PropertyGroup>
       <OriginalDotnetRootValue>$(DOTNET_ROOT)</OriginalDotnetRootValue>
-      <DOTNET_ROOT>$(RepoRoot)</DOTNET_ROOT>
+      <DOTNET_ROOT>$([System.IO.Path]::Combine('$(RepoRoot)', '.dotnet'))</DOTNET_ROOT>
```
This isn't reliable. The dotnet.cmd tool will not always point at the .dotnet directory.  Instead you need to read the directory from artifacts/toolset/sdk.txt.

trylek replied:
Fixed in 4th commit. I have also added a temporary instrumentation showing the <code>DOTNET_ROOT</code> location as the installer builds for some reason still fail to find the correct SDK even though I believe lab machines use the default <code>.dotnet</code> location.

davidwrighton reviewed (changes requested):
Correct calc of this needs to use the computed sdk path, not assume its in .dotnet. This will break developers which install a global build of the sdk which has a sufficiently matching dotnet version. (Most of the library developers do this.)

trylek committed:
Address David's PR feedback; add instrumentation to show the .NET root

AntonLapounov commented on src/installer/pkg/sfx/Microsoft.NETCore.App/ReadyToRun.targets:
```diff
@@ -46,10 +47,16 @@
 
   <Target Name="PatchDotnetRootBeforeRunningCrossgen2" BeforeTargets="_CreateR2RImages">
+    <ReadLinesFromFile File="$(ArtifactsDir)/toolset/sdk.txt">
+      <Output TaskParameter="Lines" ItemName="DotNetRoot"/>
```
```suggestion
      <Output TaskParameter="Lines" PropertyName="DOTNET_ROOT"/>
```

AntonLapounov replied:
You may assign directly to a property and remove `<DOTNET_ROOT>@(DotNetRoot)</DOTNET_ROOT>`.

trylek replied:
Thanks, will do. Sadly this doesn't change the fact that it still doesn't work, either the file doesn't exist or the script is unable to read it. I'm waiting on the Linux runs to see whether it's just a problem of the directory separator formatting, it's actually pretty hard for me to test locally as I naturally have the latest dogfooding SDK installed, otherwise I basically wouldn't be able to debug anything in the VS.

trylek committed:
Address additional Anton's PR feedback; use RepoRoot instead of ArtifactsDir

trylek committed:
Actually set the DOTNET_ROOT environment variable for Crossgen2

trylek committed:
Instrumentation

trylek committed:
Normalize path to the sdk.txt file

AntonLapounov commented on src/installer/pkg/sfx/Microsoft.NETCore.App/ReadyToRun.targets:
```diff
@@ -40,22 +41,48 @@
+    <Message Importance="High" Text="DotnetSdkRootFile: $(DotnetSdkRootFile)" />
+    <Message Importance="High" Text="DotnetRoot:        $(DotnetRoot)" />
+
+    <SetEnvVar Name="DOTNET_ROOT" Value="$(DotnetRoot)" />
```
Is this a temporary change for investigation?

trylek replied:
Yes, I keep hitting issues in properly reading the DOTNET_ROOT from the sdk.txt file, I'll delete these once the change works.

AntonLapounov replied:
I see that this change has been merged rather than removed. Why could not we use the `DotNetHostPath` metadata that you added to SDK recently?

trylek replied:
Hi Anton! Can you please clarify your comment? For the "Message" tags, I removed those. For DOTNET_ROOT, that is still necessary because technically the runtime repo still supports SDK preview 1 as the minimum i.e. the one that doesn't yet have my DotNetHostPath support. That will happen as part of the April infra rollout (I think 4/5 is the date) after which I'll be able to merge in my change

https://github.com/dotnet/runtime/pull/49004

cleaning up the hack (modulo the fact that I need to rebase the change after this check-in).

AntonLapounov replied:
My comment related to the `SetEnvVar` line, not the `Message` ones, which GitHub displays as the context. Thank you for the clarification!

trylek committed:
More instrumentation as I still fail to grasp why it doesn't work

trylek committed:
Explicitly initialize the .NET CLI per Jeremy's suggestion

trylek committed:
Simplify DOTNET_ROOT logic based on Jeremy's advice

trylek committed:
Remove instrumentation as the change finally works

trylek commented:
@davidwrighton - I believe I have addressed your PR feedback and in fact, thanks to Jeremy's advice it's ultimately super simple. I have also analyzed the binlogs on Windows (there are no binlogs produced for Linux / OSX runs, I'll create a tracking issue for that) and I confirmed that <code>RunReadyToRunCompilation</code> now truly uses Crossgen2. Can you please take another look when you have a chance?
